src/sql/rewrite/ob_transform_simplify_expr.cpp

```diff
@@ -2517,4 +2526,681 @@ int ObTransformSimplifyExpr::try_remove_ora_decode(ObRawExpr *&expr,
+        ObRawExpr *lnnvl_when_expr = NULL;
+        if (OB_ISNULL(when_expr = case_expr->get_when_param_expr(i))) {
+          ret = OB_ERR_UNEXPECTED;
+          LOG_WARN("unexpected when is NULL", K(ret));
+        } else if (OB_FAIL(ObRawExprUtils::build_lnnvl_expr(*(ctx_->expr_factory_), 
+                                                            when_expr,
+                                                            lnnvl_when_expr))) {
+          LOG_WARN("failed to build lnnvl expr", K(ret));
+        } else if (OB_FAIL(build_exprs.push_back(lnnvl_when_expr))) {
+          LOG_WARN("failed to push back build_exprs", K(ret));
+        }
+      }
+      // push back condk, if k != n
+      if (OB_SUCC(ret) && true_null_uncalc_idx < case_expr->get_then_expr_size()) {
+        ObRawExpr *when_expr = NULL;
+        if (OB_ISNULL(when_expr = case_expr->get_when_param_expr(true_null_uncalc_idx))) {
+          ret = OB_ERR_UNEXPECTED;
+          LOG_WARN("unexpected when is NULL", K(ret));
+        } else if (OB_FAIL(build_exprs.push_back(when_expr))) {
+          LOG_WARN("failed to push back build_exprs", K(ret));
+        }
+      }
+      // push back true_null_non_static expr
+      // build rewrite_expr to lnnvl(cond1) and ... and lnnvl(condk-1) and condk and exprk
+      if (OB_FAIL(ret)) {
+      } else if (OB_FAIL(build_exprs.push_back(true_null_uncalc_expr))) {
+        LOG_WARN("failed to push back build_exprs", K(ret));
+      } else if (OB_FAIL(ObRawExprUtils::build_and_expr(*ctx_->expr_factory_, 
+                                                        build_exprs, rewrite_expr))) {
+        LOG_WARN("failed to build and exprs", K(ret));
+      }
+    }
+  } /* endif build rewrite_expr */
+  return ret;
+}
+
+int ObTransformSimplifyExpr::add_false_or_false_constraint(
+                          const ObIArray<ObRawExpr *> &false_exprs) {
+  int ret = OB_SUCCESS;
+  ObRawExpr *false_cons_expr = NULL;
+  if (OB_ISNULL(ctx_)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("unexpected ctx_ is NULL", K(ret));
+  } else if (OB_LIKELY(false_exprs.count() == 0)) {
+  } else if (OB_FAIL(ObRawExprUtils::build_or_exprs(*ctx_->expr_factory_, 
+                                                    false_exprs, false_cons_expr))) {
+    LOG_WARN("failed to build or expr", K(ret));
+  } else if (OB_ISNULL(false_cons_expr)) {
+    ret = OB_ERR_UNEXPECTED;
+    LOG_WARN("unexpected null expr", K(ret));
+  } else if (OB_FAIL(false_cons_expr->formalize(ctx_->session_info_))) {
+    LOG_WARN("failed to formalize expr", K(ret), K(false_cons_expr));
+  } else {
+    ObExprConstraint false_cons(false_cons_expr, PreCalcExprExpectResult::PRE_CALC_RESULT_FALSE);
+    if (OB_FAIL(ctx_->expr_constraints_.push_back(false_cons))) {
+      LOG_WARN("failed to push back constraints");
+    }
+  }
+  return ret;
+}
+
+int ObTransformSimplifyExpr::add_true_and_true_constraint(
+                          const ObIArray<ObRawExpr *> &true_exprs) {
```

Comment:
remove this unused function: ObTransformSimplifyExpr::add_true_and_true_constraint